src/handlers/api.ts

```diff
@@ -0,0 +1,204 @@
+    const locks: Lock[] = [];
+    locksMap.forEach((v, k) => locks.push({ name: k, owner: v } as Lock));
+    console.log("Retrieved locks", { locks });
+    res.status(200).json(locks);
+  }
+);
+
+app.get(
+  "/api/teams/:team/channels/:channel/locks/:lock",
+  authorizer,
+  async (req, res) => {
+    const { team, channel, lock: lockName } = req.params;
+    const lockOwner = await lockRepo.getOwner(lockName, channel, team);
+    if (lockOwner) {
+      const lock = { name: lockName, owner: lockOwner } as Lock;
+      console.log("Retrieved lock", { lock });
+      res.status(200).json(lock);
+    } else {
+      console.log("Lock not found", { lockName });
+      res.status(404).json({ error: `${lockName} not found` });
+    }
+  }
+);
+
+const bodyValidator = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  const decoded = Lock.decode(req.body);
+  if (isLeft(decoded)) {
+    const error = D.draw(decoded.left);
+    console.log("Invalid request body", { body: req.body, error });
+    res.status(400).json({ error });
+  } else {
+    next();
+  }
+};
+
+app.post(
+  "/api/teams/:team/channels/:channel/locks",
+  authorizer,
+  bodyValidator,
+  async (req, res) => {
+    const username = auth(req)!.name;
+    const { channel, team } = req.params;
+    const lock = req.body as Lock;
+    if (lock.owner !== username) {
+      const error = `${username} cannot lock for another user ${lock.owner}`;
+      console.log("Cannot lock", { username, body: req.body, error });
+      res.status(403).json({ error });
+    } else {
+      const lockOwner = await lockRepo.getOwner(lock.name, channel, team);
+      if (!lockOwner) {
+        await lockRepo.setOwner(lock.name, lock.owner, channel, team);
+        console.log("Added lock", { lock });
+        res.status(201).json(lock);
+      } else if (lockOwner === lock.owner) {
+        console.log("Lock exists", { lock });
+        res.status(200).json(lock);
+      } else {
+        const error = `${lock.name} is already locked by ${lockOwner}`;
+        console.log("Cannot lock", { lock, lockOwner, error });
+        res.status(403).json({ error });
+      }
+    }
+  }
+);
+
+const paramsValidator = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  const Params = D.type({
+    lock: NonEmptyWhitespaceFreeString,
+  });
+  const decoded = Params.decode(req.params);
+  if (isLeft(decoded)) {
+    const error = D.draw(decoded.left);
+    console.log("Invalid request params", { params: req.params, error });
+    res.status(400).json({ error });
+  } else {
+    next();
+  }
+};
+
+app.delete(
+  "/api/teams/:team/channels/:channel/locks/:lock",
+  authorizer,
+  paramsValidator,
+  async (req, res) => {
+    const { team, channel, lock: lockName } = req.params;
+    const lockOwner = await lockRepo.getOwner(lockName, channel, team);
+    if (!lockOwner) {
+      console.log("No lock to delete", { lockName });
+      res.status(204).end();
```

Comment:
Would returning a 410 Gone error work here?

Reply:
Nice spot 🧐

`410` is an interesting one, I've never used it! It might not be quite accurate all the time because can something be `Gone` if it was never there in the first place? Or only `Gone` if it was previously locked?

This was one where I wasn't quite sure what to return and opted for a `2XX`/success code over `4XX` because I assumed that clients wouldn't care (or want an error) for when they're trying to delete a lock and there isn't one. But my assumption might be very wrong.